posts/2018-06-15-typedapi.md

```diff
@@ -0,0 +1,329 @@
+
+## Servant in a nutshell and how it began
+For everyone not knowing Servant, it is a library which lets you define your web apis as types and derives the client and server functions from it. When I saw it for the first time while working on a ped project I immediately loved the idea. Creating web server and clients this way reduces your code to a mere type, you get extra type safety and you can use the api types as contracts between your server and its clients. 
+
+I couldn't find any viable alternative in Scala at the time and decided to build it on my own. But I just wanted to start with a single feature to not overwhelm myself and abandon the project after a short time. Therefore, I set out to make Scala able to derive a client function from a single api type, as we will do in this post.
+
+## Derive a client function from a type. How hard can it be?
+Let's start with an example we will use later on to ease understanding. Consider the following api:
+
+```
+GET /users/:name?minAge=:age -> List[User]
+```
+
+It only consists of a single endpoint which returns a list of `Users`:
+
+```scala
+final case class User(name: String, age: Int)
+```
+
+with a given `name: String`. Furthermore, you filter the resulting users by their `age: Int`. Our big goal is to end up with a function which is derived from a type-level representation of our endpoint:
+
+```scala
+(name: String, minAge: Int) => F[List[User]]
+```
+
+### Represent the api as a type
+Question: how do you represent the above api as a type in Scala? I think the best way is to break it apart and try to find type-level representations for each element. After that, we "just" merge them together.
+
+When we take a closer look at our endpoint we see that it consists of:
+  * path elements identifying an endpoint: `/users`
+  * segments within the path: `:name`
+  * queries: `minAge=[age]`
+  * and a method `GET` to identify which kind of operation we want to do
+
+Or in other words, just a plain HTTP definition of a web endpoint. Now that we know what we are working with let's try and find a type-level representation.
+
+We start with our path element. If we would work with Dotty or Typelevel-Scala we could use a feature called literal-type:
+
+```scala
+type Path = "users"
+```
+
+But as we want to stay in Vanilla Scala this isn't an option.
+We have to use the one tool probably every developer has to use when it comes to working on the type-level called [shapeless](https://github.com/milessabine/shapeless).
+It has this nifty class [Witness](https://github.com/milessabin/shapeless/blob/shapeless-2.3.3/core/src/main/scala/shapeless/singletons.scala#L32) which generates a witness type from a literal like `Strings` or `Symbols`.
+
+```scala
+import shapeless.Witness
+
+val usersW = Witness("users")
+```
+
+But this isn't a pure type declaration, you will say. And you are right, but right now there is no other way in Scala. We have to go the ordinary value road first to create our types.
```

Comment:
If I didn't knew about `Witness` and singleton types, this wouldn't be very clear to me. Maybe it would be a little clearer it you flipped the reasonning like so: 
* the challenge here is to "transform a value into a type" 
* of course we can do that only if the value is already known at compile time (eg a literal)
* Dotty can do exactly that, but in Scala we need to use `Witness`
* a `Witness` *value* "contains" an abstract type `T` that is exactly what we want here

Reply:
True, when you use a concept on a (more or less) daily basis you easily forget that it might be not that obvious for someone else why it should/can be used here. 

I will rewrite it to make it more clear.